rules.md

```diff
@@ -0,0 +1,18 @@
+# Substrate
+
+- 🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/polkadot-review.
```

Comment:
It seems like there's some disagreement about this requirement and others similar to it. I'll try to decompose the sentence according to my literal interpretation:

> 🔒️ed lines are reviewed by **2** team leads

Emphasis on "2".

> one of them must be in ...

Let's call this agent `DeveloperA`.

> and the other in ...

"the other" literally refers to another agent different from `DeveloperA`. Let's call this `DeveloperB`.

All of the above phrasing to me strongly suggests 2 different agents for this requirement.

Now it might be that someone is in both `@paritytech/locks-review` and `@paritytech/polkadot-review`. https://matrix.to/#/!gJeGMHCcDoIwsHIJri:matrix.parity.io/$5H7WC-VGc4Am5HuvivqpuqA-jR_ic2SmWiz0x4HK1cg raised the question:

> why are you only counting towards one group if you are part of multiple? 

If we'd make it so 1 person can fulfill the criteria by being in both groups, we'd effectively lower the requirement to a single approval (1 person) instead of 2 like I understood by the literal interpretation. If that's intentional, I suggest clarifying the phrasing as e.g.

>🔒️ed requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review

That removes any implication of "2 team leads" since it only mentions 1 approval from each group, which might be given by the same person if they are in both groups. What's your opinion?

cc @bkchr @TriplEight

Reply:
I agree that 1 person that is a member of two teams shall count as a review from both teams. But also it should count as **one** review of two requested.

Reply:
> A second team lead from either team is then required to make the check succeed.

"Either" team? That's different from how I was understanding your suggestion.

I think it'll help to walk through an example scenario. Suppose the following structure

```
├── locks-review
│  ├── user1
│  └── user2
├── polkadot-review
│  ├── user1
│  └── user3
```

**Alternative 1** (how it works currently): after `user1` approves we'll still need an approval from `user2` or `user3`, even though `user1` is in both teams.

Suggested wording:

> :lock: lines requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review. Each approval should come from a different user.

Note: I think adding "Each approval should come from a different user" is clearer than saying "2 team leads".

---

**Alternative 2** (what I _**initially**_ understood from your suggestion): `user1`'s approval would fulfill the requirement because they are on both teams, therefore not requiring an approval from `user2` or `user3`.

Suggested wording:

> :lock: lines requires 1 approval from @paritytech/locks-review and 1 approval from @paritytech/polkadot-review

Note: avoids mentioning "2 team leads" in order to not provoke any ambiguity, since a single team lead is enough to fulfill both requirements. Also avoids mentioning "Each approval should come from a different user" since it's intentionally not that way.

---

**Alternative 3** (how I understood from "either"): after `user1` approves we'll need an approval from **either** `user2` or `user3`.

Suggested wording:

> :lock: lines requires 2 approvals from users on either @paritytech/locks-review or @paritytech/polkadot-review. Each approval should come from a different user.

---

Which alternative are you suggesting? Also feel free to describe another alternative if none is matching

Reply:
Isn't alternative 1 and 3 the same? 

And is alternative 1 really how it currently works? I doubt this.

Reply:
Okay, I think I may mixed something up. I'm now confused what the current status is. I just know that yesterday it wasn't alternative 1.

Reply:
I have now extended the polkadot-review team and now it should be okay.

Reply:
> Isn't alternative 1 and 3 the same?

In this specific example, when `user1` approves, they *result* in the same outcome but their definitions are still  different.

Consider if `user2` were to approve. "Alternative 1" in that case would require that other approval comes from specifically `polkadot-review` (`user2` isn't a part of it). However, if you say "total approvals from either team" then the second approval could come from the same team that `user2` is in. In this reduced example with a few team members they would might effectively result in the same conditions, but I think it would be different if you start adding more team members like the groups we have currently.

> I just know that yesterday it wasn't alternative 1

If it's about https://github.com/paritytech/polkadot/runs/5978667098?check_suite_focus=true#step:2:43 it was misbehaving due to a bug, but it should be fine now.

Reply:
Then let's take alternative 1

Reply:
FWIW I thought that the initial "🔒️ed lines are reviewed by 2 team leads, one of them must be in @paritytech/locks-review and the other in @paritytech/polkadot-review." was precise enough to imply alternative 1 as-is, but if that is unclear, I'd go for something like:

> 🔒️ed lines must be reviewed by 2 team leads. One of them must be a member of @paritytech/locks-review and the other must be a member of @paritytech/polkadot-review.

If that is _still_ unclear, then I'd add somehting like @bkchr said:

> 🔒️ed lines must be reviewed by 2 team leads. One of them must be a member of @paritytech/locks-review and the other must be a member of @paritytech/polkadot-review. If one is a member of both teams, then the other may be a member of either team.